File: submit-api/src/submit_api/models/package.py
"""Submission package model class.

Manages the package
"""
from __future__ import annotations

import enum

from sqlalchemy import Column, Enum, ForeignKey
from sqlalchemy.orm import joinedload

from .package_version import PackageVersion
from .base_model import BaseModel
from .db import db


class NonCanonicalPackageStatus(enum.Enum):
    """Enum for non-canonical package statuses."""

    UPDATED = "UPDATED"
    UPDATE_REQUESTED = "UPDATE_REQUESTED"
    REVISION_REQUESTED = "REVISION_REQUESTED"
    REVISION_REQUIRED = "REVISION_REQUIRED"

    @classmethod
    def check_value(cls, value):
        """Check if a value is a valid non-canonical package status."""
        for member in cls:
            if member.value == value:
                return member
        return None


class PackageStatus(enum.Enum):
    """Enum for package statuses."""

    IN_REVIEW = 'IN_REVIEW'
    APPROVED = 'APPROVED'
    ACCEPTED = 'ACCEPTED'
    SATISFIED = 'SATISFIED'
    REJECTED = 'REJECTED'
    REVIEWED = 'REVIEWED'
    SUBMITTED = 'SUBMITTED'
    PARTIALLY_COMPLETED = 'PARTIALLY_COMPLETED'
    COMPLETED = 'COMPLETED'
    NEW_SUBMISSION = 'NEW_SUBMISSION'
    NEW = 'NEW'
    IN_PROGRESS = 'IN_PROGRESS'
    PASSED_CONSULTATION_CHECK = 'PASSED_CONSULTATION_CHECK'
    FAILED_CONSULTATION_CHECK = 'FAILED_CONSULTATION_CHECK'
    UNDER_REVIEW = 'UNDER_REVIEW'
    UNDER_CONSULTATION_CHECK = 'UNDER_CONSULTATION_CHECK'
    REVIEW_REJECTED = 'REVIEW_REJECTED'
    REVIEW_NOT_COMPLETED = 'REVIEW_NOT_COMPLETED'
    CREATED = 'CREATED'
    AWAITING_MANAGER_APPROVAL = 'AWAITING_MANAGER_APPROVAL'
    CC_AWAITING_MANAGER_APPROVAL = 'CC_AWAITING_MANAGER_APPROVAL'
    MP_AWAITING_MANAGER_APPROVAL = 'MP_AWAITING_MANAGER_APPROVAL'
    IEM_AWAITING_MANAGER_APPROVAL = 'IEM_AWAITING_MANAGER_APPROVAL'
    REVISION_REQUIRED = 'REVISION_REQUIRED'
    NO_REVISION_REQUIRED = 'NO_REVISION_REQUIRED'
    RESUBMITTED = 'RESUBMITTED'
    REQUESTED_BY_EAO = 'REQUESTED_BY_EAO'

    @classmethod
    def check_value(cls, value):
        """Check if a value is a valid package status."""
        for member in cls:
            if member.value == value:
                return member
        return None


class Package(BaseModel):
    """Definition of the package entity."""

    __tablename__ = 'packages'

    id = Column(db.Integer, primary_key=True, autoincrement=True)
    account_project_id = Column(db.Integer, ForeignKey(
        'account_projects.id', ondelete='CASCADE'), nullable=False)
    name = Column(db.String(255), nullable=False)
    description = Column(db.String(500), nullable=True, comment="Description of the package")
    type_id = Column(db.Integer, ForeignKey(
        'package_types.id'), nullable=False)
    type = db.relationship('PackageType', foreign_keys=[
                           type_id], lazy='joined')
    account_project_work_id = Column(db.Integer, ForeignKey(
        'account_project_works.id', ondelete='SET NULL'), nullable=True)
    account_project_work = db.relationship('AccountProjectWork', foreign_keys=[
                           account_project_work_id], lazy='joined', back_populates='packages')
    submitted_on = Column(db.DateTime, nullable=True)
    submitted_by = Column(db.String, ForeignKey(
        'users.auth_guid', name='packages_submitted_by_fkey'), nullable=True)
    submitted_by_user = db.relationship(
        'User', foreign_keys=[submitted_by], lazy='joined')
    completed_on = Column(db.DateTime, nullable=True)
    meta = db.relationship(
        'PackageMetadata', backref='package', lazy='joined', uselist=False, cascade='all, delete', passive_deletes=True)
    items = db.relationship('Item', backref='package',
                            lazy='select', order_by='Item.sort_order', cascade='all, delete', passive_deletes=True)
    status = Column(db.ARRAY(Enum(PackageStatus)), nullable=False,
                    default=[PackageStatus.NEW.value])
    active = Column(db.Boolean, nullable=False, default=True)
    version_id = Column(db.Integer, ForeignKey(
        'package_versions.id'), nullable=True)
    version = db.relationship('PackageVersion', foreign_keys=[
                              version_id], lazy='joined')

    __table_args__ = (
        db.Index('idx_packages_account_project_id', 'account_project_id'),
        db.Index('idx_packages_account_project_work_id', 'account_project_work_id',
                 postgresql_where=db.text('account_project_work_id IS NOT NULL')),
    )

    _update_requests = db.relationship(
        'UpdateRequest',
        backref='package',
        lazy='joined')

    internal_staff_documents = db.relationship('InternalStaffDocument',
                                               backref='package',
                                               lazy='joined',
                                               cascade='all, delete',
                                               passive_deletes=True)

    @property
    def update_requests(self):
        """Get the active update requests for the package."""
        return [ur for ur in self._update_requests if ur.active]

    @property
    def all_update_requests(self):
        """Get all update requests for the package."""
        return self._update_requests

    @classmethod
    def get_package_by_id_with_items(cls, package_id: int):
        """Return model by package id."""
        return cls.query.filter_by(id=package_id).options(joinedload(Package.items)).first()

    @classmethod
    def get_all_package_by_ids(cls, package_ids: list[int]):
        """Return model by package ids."""
        return cls.query.filter(Package.id.in_(package_ids)).all()

    @classmethod
    def get_all_latest_packages_by_original_package_ids(cls, original_package_ids: list[int]):
        """Return all packages with the greatest PackageVersion.version by original package ids."""
        subquery = (db.session.query(
            PackageVersion.original_package_id,
            db.func.max(PackageVersion.version).label('max_version')
        ).filter(PackageVersion.original_package_id.in_(original_package_ids))
                    .group_by(PackageVersion.original_package_id).subquery())

        return cls.query.join(PackageVersion).join(
            subquery,
            (PackageVersion.original_package_id == subquery.c.original_package_id) &
            (PackageVersion.version == subquery.c.max_version)
        ).all()

    @classmethod
    def get_account_project_id_by_package_id(cls, package_id: int) -> int | None:
        """Return only the account_project_id for a given package id."""
        result = db.session.query(cls.account_project_id).filter(cls.id == package_id).scalar()
        return result

File: submit-web/src/models/Package.ts
import { InternalStaffDocument, SubmissionItem } from "./SubmissionItem";
import { UpdateRequest } from "./UpdateRequest";
import { AccountProjectWork } from "./AccountProjectWork";
import { StatusEntry, buildRoleFilters } from "./Status";

export enum SubmissionPackageType {
  MANAGEMENT_PLAN = "Management Plan",
  IEM = "IEM",
  IPD = "IPD",
  ADDITIONAL_INFORMATION = "Additional Information",
}

export const NON_WITHDRAWABLE_UPDATE_REQUEST_PACKAGE_TYPES: SubmissionPackageType[] =
  [SubmissionPackageType.MANAGEMENT_PLAN, SubmissionPackageType.IEM];

export enum SubmissionPackageApprovalType {
  A = "A",
  B = "B",
  C = "C",
}

export type PackageType = {
  id: number;
  name: SubmissionPackageType;
  title?: string;
  versioning_enabled: boolean;
  mandatory: boolean;
  approval_type?: SubmissionPackageApprovalType;
  success_message?: string;
};

// These statuses are just for UI purposes, the actual canonical business statuses are PackageStatus
export type NonCanonicalPackageStatus =
  | "REVISION_REQUIRED"
  | "UPDATED"
  | "NO_REVISION_REQUIRED"
  | "RESUBMITTED"
  | "UPDATE_REQUESTED"
  | "REVISION_REQUESTED";

export const NON_CANONICAL_PACKAGE_STATUS = Object.freeze<
  Record<NonCanonicalPackageStatus, NonCanonicalPackageStatus>
>({
  UPDATE_REQUESTED: "UPDATE_REQUESTED",
  REVISION_REQUIRED: "REVISION_REQUIRED",
  REVISION_REQUESTED: "REVISION_REQUESTED",
  UPDATED: "UPDATED",
  NO_REVISION_REQUIRED: "NO_REVISION_REQUIRED",
  RESUBMITTED: "RESUBMITTED",
});

export type PackageStatus =
  | "IN_REVIEW"
  | "UNDER_REVIEW"
  | "APPROVED"
  | "ACCEPTED"
  | "SATISFIED"
  | "REJECTED"
  | "REVIEW_REJECTED"
  | "REVIEW_NOT_COMPLETED"
  | "REVIEWED"
  | "COMPLETED"
  | "SUBMITTED"
  | "PARTIALLY_COMPLETED"
  | "NEW_SUBMISSION"
  | "NEW"
  | "IN_PROGRESS"
  | "UNDER_CONSULTATION_CHECK"
  | "PASSED_CONSULTATION_CHECK"
  | "AWAITING_MANAGER_APPROVAL"
  | "FAILED_CONSULTATION_CHECK"
  | "CREATED"
  | "REQUESTED_BY_EAO"
  | "INTERNAL_VERIFICATION"
  | "VERIFIED"
  | "PENDING_ACKNOWLEDGEMENT"
  | "READY_FOR_ACKNOWLEDGEMENT"
  | "READY_FOR_APPROVAL"
  | "ACKNOWLEDGED"
  | "NOT_APPROVED"
  | "WITHDRAWN";

export const PACKAGE_STATUS: Record<
  PackageStatus,
  StatusEntry<PackageStatus | NonCanonicalPackageStatus>
> = {
  IN_REVIEW: {
    value: "IN_REVIEW",
    label: "In Review",
  },
  UNDER_REVIEW: {
    value: "UNDER_REVIEW",
    label: "Under Review",
    sortOrder: 4,
  },
  APPROVED: {
    value: "APPROVED",
    label: "Approved",
    sortOrder: 13,
  },
  REJECTED: {
    value: "REJECTED",
    label: "Rejected",
  },
  REVIEW_REJECTED: {
    value: "REVIEW_REJECTED",
    label: "Review Rejected",
  },
  REVIEW_NOT_COMPLETED: {
    value: "REVIEW_NOT_COMPLETED",
    label: "Review Not Completed",
  },
  COMPLETED: {
    value: "COMPLETED",
    label: "Completed",
    sortOrder: 7,
  },
  SUBMITTED: {
    value: "SUBMITTED",
    label: "Submitted",
    filter: ["proponent"],
    sortOrder: 3,
  },
  PARTIALLY_COMPLETED: {
    value: "PARTIALLY_COMPLETED",
    label: "Partially Completed",
    sortOrder: 5,
  },
  NEW: {
    value: "NEW",
    label: "New",
    filter: ["proponent"],
    sortOrder: 1,
  },
  IN_PROGRESS: {
    value: "IN_PROGRESS",
    label: "In Progress",
    filter: ["proponent"],
    sortOrder: 2,
  },
  NEW_SUBMISSION: {
    value: "NEW_SUBMISSION",
    label: "New Submission",
    sortOrder: 1,
  },
  UNDER_CONSULTATION_CHECK: {
    value: "UNDER_CONSULTATION_CHECK",
    label: "Under Consultation Check",
    filter: ["proponent"],
    sortOrder: 4,
  },
  PASSED_CONSULTATION_CHECK: {
    value: "PASSED_CONSULTATION_CHECK",
    label: "Passed Consultation Check",
    filter: ["proponent"],
    sortOrder: 5,
  },
  CREATED: {
    value: "CREATED",
    label: "Created",
  },
  ACCEPTED: {
    value: "ACCEPTED",
    label: "Accepted",
    sortOrder: 11,
  },
  SATISFIED: {
    value: "SATISFIED",
    label: "Satisfied",
    sortOrder: 12,
  },
  AWAITING_MANAGER_APPROVAL: {
    value: "AWAITING_MANAGER_APPROVAL",
    label: "Awaiting Manager Approval",
    sortOrder: 7,
  },
  FAILED_CONSULTATION_CHECK: {
    value: "FAILED_CONSULTATION_CHECK",
    label: "Failed Consultation Check",
    sortOrder: 8,
  },
  REVIEWED: {
    value: "REVIEWED",
    label: "Reviewed",
    sortOrder: 10,
  },
  INTERNAL_VERIFICATION: {
    value: "INTERNAL_VERIFICATION",
    label: "Internal Verification",
  },
  VERIFIED: {
    value: "VERIFIED",
    label: "Verified",
  },
  REQUESTED_BY_EAO: {
    value: "REQUESTED_BY_EAO",
    label: "Requested by EAO",
  },
  PENDING_ACKNOWLEDGEMENT: {
    value: "PENDING_ACKNOWLEDGEMENT",
    label: "Pending Acknowledgement",
  },
  READY_FOR_ACKNOWLEDGEMENT: {
    value: "READY_FOR_ACKNOWLEDGEMENT",
    label: "Ready for Acknowledgement",
  },
  ACKNOWLEDGED: {
    value: "ACKNOWLEDGED",
    label: "Acknowledged",
  },
  READY_FOR_APPROVAL: {
    value: "READY_FOR_APPROVAL",
    label: "Ready for Approval",
  },
  NOT_APPROVED: {
    value: "NOT_APPROVED",
    label: "Not Approved",
  },
  WITHDRAWN: {
    value: "WITHDRAWN",
    label: "Withdrawn",
  },
};

export const EAO_PACKAGE_STATUS_FILTERS = buildRoleFilters(
  PACKAGE_STATUS,
  "eao",
);

export const PROPONENT_PACKAGE_STATUS_FILTERS = buildRoleFilters(
  PACKAGE_STATUS,
  "proponent",
);

export type SubmissionPackageMeta = Record<string, any>;

export type PackageVersion = {
  id: number;
  package_id: number;
  version: number;
  original_package_id: number;
  is_approved: boolean;
  is_enforceable: boolean;
  is_latest: boolean;
};

export type SubmissionPackage = {
  id: number;
  name: string;
  description?: string;
  status: PackageStatus[];
  submitted_on?: string;
  completed_on?: string;
  submitted_by?: string;
  type_id: number;
  type: PackageType;
  items: Array<SubmissionItem>;
  account_project_id: number;
  account_project_work?: AccountProjectWork;
  meta?: SubmissionPackageMeta;
  days_since_submission?: number;
  internal_staff_documents?: InternalStaffDocument[];
  review_status?: NonCanonicalPackageStatus;
  update_requests: Array<UpdateRequest>;
  all_update_requests: Array<UpdateRequest>;
  version: PackageVersion;
  enforceable?: boolean;
};
